File: 2022-05/typescript/234_Palindrome_Linked_List.ts
function isPalindrome(head: null | ListNode): boolean {
    let rev: null | ListNode = null
    let slow = head
    let fast = head
    while (fast && fast.next) {
        fast = fast.next.next
        let tmp = slow
        slow = slow.next
        tmp.next = rev
        rev = tmp
    }
    if (fast) {
        slow = slow.next
    }
    while (rev && rev.val === slow.val) {
        rev = rev.next
        slow = slow.next
    }
    return !rev
}

File: 2022-05/python/234_Palindrome_Linked_List.py
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        rev_half = None
        slow, fast = head, head
        while fast and fast.next:
            fast = fast.next.next
            rev_half, rev_half.next, slow = slow, rev_half, slow.next
        if fast:
            slow = slow.next
        while rev_half and rev_half.val == slow.val:
            rev_half = rev_half.next
            slow = slow.next
        return not rev_half
